### ISSUE TITLE
Cannot Start B-Hyve Zone from Homekit

### ISSUE THREAD

**Issue by mitch7391** [codyc1515/homebridge-orbit-bhyve#12]
Hey @codyc1515 I know you’re a busy man but thought I’d put this one in for when you get time. My b-hyve is not ever doing anything when trying to action a station to start. When turned on it just saying ‘waiting’ and nothing happens, then I turn it off. The Homebridge logs never indicate that it attempting anything, I don’t see a story command issued or anything. I do see the ‘logged in’ and ‘battery service’ log entries, so I at least know it’s receiving something. Once and only once did it attempt to start and it killed my HomeKit with the follow Homebridge logs. I have not been able to get it to go this far again, it just does nothing. The following was an attempt to start zone 1.

Oct 23 18:50:40 homebridge homebridge[10651]: [10/23/2019, 6:50:40 PM] Error: WebSocket is not open: readyState 0 (CONNECTING)
Oct 23 18:50:40 homebridge homebridge[10651]: at WebSocket.send (/usr/local/lib/node_modules/homebridge-orbit-bhyve/node_modules/ws/lib/websocket.js:327:13)
Oct 23 18:50:40 homebridge homebridge[10651]: at BHyveValve.open (/usr/local/lib/node_modules/homebridge-orbit-bhyve/index.js:501:24)
Oct 23 18:50:40 homebridge homebridge[10651]: at WebSocket.emit (events.js:193:13)
Oct 23 18:50:40 homebridge homebridge[10651]: at WebSocket.setSocket (/usr/local/lib/node_modules/homebridge-orbit-bhyve/node_modules/ws/lib/websocket.js:172:10)
Oct 23 18:50:40 homebridge homebridge[10651]: at ClientRequest.req.on (/usr/local/lib/node_modules/homebridge-orbit-bhyve/node_modules/ws/lib/websocket.js:646:15)
Oct 23 18:50:40 homebridge homebridge[10651]: at ClientRequest.emit (events.js:193:13)
Oct 23 18:50:40 homebridge homebridge[10651]: at TLSSocket.socketOnData (_http_client.js:480:11)
Oct 23 18:50:40 homebridge homebridge[10651]: at TLSSocket.emit (events.js:193:13)
Oct 23 18:50:40 homebridge homebridge[10651]: at addChunk (_stream_readable.js:295:12)
Oct 23 18:50:40 homebridge homebridge[10651]: at readableAddChunk (_stream_readable.js:276:11)
Oct 23 18:50:40 homebridge homebridge[10651]: at TLSSocket.Readable.push (_stream_readable.js:231:10)
Oct 23 18:50:40 homebridge homebridge[10651]: at TLSWrap.onStreamRead (internal/stream_base_commons.js:154:17)
Oct 23 18:50:40 homebridge homebridge[10651]: [10/23/2019, 6:50:40 PM] Got SIGTERM, shutting down Homebridge...

**Comment by mitch7391** [comment]
I just saw that there was a new update and updated. Once up and running I attempted to turn on zone 3 and got the above error for the second time, but once again cannot produce it again. Both times I have come close to executing a start command has only been after an update and then any time after there is nothing in the homebridge logs and in homekit it just sits there saying 'waiting'.

I seem to have some issue with the set commands, I can confirm the get commands are working as I ran my sprinkles from the B-hyve app and it reflected correctly in homekit (duration and all).  Any idea what I could be missing or doing wrong here? I had Orbit update my timer's firmware, is there a particular version of npm, node or homebridge this requires maybe? Just trying to rule out the easy stuff.

**Comment by mitch7391** [comment]
Further testing:

I would start a zone from the B-hyve app (seen as I cant from homekit), the feedback in homekit would reflect correctly that it is running and the duration is correct. I would then issue a stop command from homekit and it successfully stops the zone from homekit. So my issue only seems to be with starting a zone. Also once the zone is manually stopped, the app pushes a notification to tell you it has been manually stopped, homebridge throws the follow message at this time:

Oct 27 16:09:01 homebridge homebridge[7369]: [10/27/2019, 4:09:01 PM] [Orbit B-Hyve] WS | Unknown WS message received
Oct 27 16:09:01 homebridge homebridge[7369]: [10/27/2019, 4:09:01 PM] [Orbit B-Hyve] WS | Unknown WS message received

**Comment by mitch7391** [comment]
Might open another issue for the notifications...

**Comment by codyc1515** [comment]
So the issue is actually caused by this line: `WebSocket is not open: readyState 0 (CONNECTING)`

Now we know what the issue is we can work to fix it. I haven't found a fix for this yet but it basically means that we send the message to server before the connection has been fully established. This is because we establish a new Web Socket connection each time an action is performed and it cannot be completed in time. To me, this seems like an issue with the library _ws_. We are checking if the connection is open firstly but it tells us it's open when it's actually still connecting.

Probably the best way to sort this is to keep open the Web Socket all of the time, although the server will kill the sessions after 30-seconds if we don't go out of our way to keep it alive.

I will take a look again in the next few days.

**Comment by mitch7391** [comment]
Oh fair enough, I am glad you understand what the error is. I shall wait patiently :)

**Comment by mitch7391** [comment]
@codyc1515 have you had any time to take a look at this yet? No pressure, just checking in.

**Comment by mitch7391** [comment]
Hey @codyc1515 just checking in again :) are there any plans to jump back into this again soon?

**Comment by mitch7391** [comment]
Hey @codyc1515 do you think you might be jumping back into this work anytime soon?

**Comment by mitch7391** [comment]
Hey @codyc1515 just bumping this again :)

**Comment by sabresirish** [comment]
> Hey @codyc1515 just bumping this again :)

@mitch7391 - there is a new plugin for the bhyve by @mortjc that uses the framework developed by @codyc1515 

https://www.npmjs.com/package/homebridge-platform-orbit

**Comment by mitch7391** [comment]
@sabresirish thanks for the info! I had noticed this one but didn’t want to jump ship, but I will give it a go and see if it works for me :)

**Comment by mitch7391** [comment]
@sabresirish thanks for the heads up, just installed @MortJC’s plug-in for the first time ever I have been able to use my B-Hyve from HomeKit! It works flawlessly!

**Comment by sabresirish** [comment]
One issue I have is my bhyve does not show up as an accessory to add into scenes or automations. Do you have the same experience?

**Comment by mitch7391** [comment]
Mine shows up as an accessory in HomeKit but you cannot use the irrigation accessory for automations or scenes yet; I believe this is currently an Apple limitation.

**Comment by sabresirish** [comment]
Ok - just making sure it wasn’t an installation issue by me.

Thanks

**Comment by leehadassin** [comment]
Thanks! @MortJC plug-in works great for me. One thing to note, while apples own home app doesn’t allow the sprinklers to be added to scenes and automations, 3rd party home apps do - I just tried in Home+ and was able to set  a zone in a scene. Furthermore, one you create a scene in home+ it shows up for use in the Apple home app!!!
Home+ is not free but it is way more powerful than the standard home app and exposes way more characteristics.
